core/trino-spi/src/main/java/io/trino/spi/function/FunctionProvider.java

```diff
@@ -43,4 +46,34 @@ default TableFunctionProcessorProvider getTableFunctionProcessorProvider(Connect
     {
         throw new UnsupportedOperationException("%s does not provide table functions".formatted(getClass().getName()));
     }
+
+    default void addFunctions(FunctionBundle functionBundle)
```

Comment:
This doesn't sound right in SPI.

Before the change, the FunctionProvider was read-only (pull-based).
When implementing FunctionProvider in a plugin, i should not be asked to expose functions mutating my internal state.
This change makes it unclear what the FunctionProvider implementor is supposed to do.

Reply:
Looking at the next commit, the change is to support `Use FunctionProvider interface instead of direct binding` in trino-main. For that we would want an interface in trino-main, without any backwards compatibility guarantees SPI comes with.

Reply:
> Before the change, the FunctionProvider was read-only (pull-based).
> When implementing FunctionProvider in a plugin, i should not be asked to expose functions mutating my internal state.

Refactored, I moved only read-only (pull-based) to FunctionProvider